### PR TITLE
recording user handle in authenticator

### PR DESCRIPTION
**Pull request by variable**
This adds a `user_handle` field to authenticator so that in passwordless login, the authenticator can be looked up by `user_handle` from `navigator.credentials.get()`

This is for https://github.com/CZ-NIC/django-fido/issues/140

### CODE REVIEW

**Comment by variable** [comment]
I don't understand when I run `python manage.py test`, I get this test failed but not during `tox`, shouldn't this suppose to fail in `tox` as well?

```
FAIL: test_no_user (django_fido.tests.test_views.TestFido2AuthenticationRequestView)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jlin/projects/django-fido/django_fido/tests/test_views.py", line 210, in test_no_user
  ....
  File "/Users/jlin/projects/django-fido/django_fido/views.py", line 331, in create_fido2_request
    assert user and user.is_authenticated, "User must not be anonymous for FIDO 2 requests."
AssertionError: User must not be anonymous for FIDO 2 requests.
```

**Comment by variable** [comment]
> I don't understand when I run `python manage.py test` without using the `--settings=django_fido.tests.settings` I get the following assertion, 
> 
> ```
> FAIL: test_no_user (django_fido.tests.test_views.TestFido2AuthenticationRequestView)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/jlin/projects/django-fido/django_fido/tests/test_views.py", line 210, in test_no_user
>   ....
>   File "/Users/jlin/projects/django-fido/django_fido/views.py", line 331, in create_fido2_request
>     assert user and user.is_authenticated, "User must not be anonymous for FIDO 2 requests."
> AssertionError: User must not be anonymous for FIDO 2 requests.
> ```



> I don't understand when I run `python manage.py test` without `--settings=django_fido.test.settings` I get this test failed, what's special in the test settings that prevented this?
> 
> ```
> FAIL: test_no_user (django_fido.tests.test_views.TestFido2AuthenticationRequestView)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/jlin/projects/django-fido/django_fido/tests/test_views.py", line 210, in test_no_user
>   ....
>   File "/Users/jlin/projects/django-fido/django_fido/views.py", line 331, in create_fido2_request
>     assert user and user.is_authenticated, "User must not be anonymous for FIDO 2 requests."
> AssertionError: User must not be anonymous for FIDO 2 requests.
> ```

Nevermind, found the issue,  `two_setp_auth` in the test settings is `True`

**Comment by variable** [comment]
I finally managed to get 2 yubi keys, but it seems the user ids for 2 different keys for 2 different user generate the same base64 string, it's weird, hence I am pausing this PR.

**Comment by variable** [comment]
Nevermind, just realised my browser was overwriting session for each user, doh!

File: django_fido/forms.py [hunk 1]
```diff
@@ -19,6 +19,7 @@ class Fido2RegistrationForm(forms.Form):
                                   widget=forms.HiddenInput)
     attestation = forms.CharField(error_messages={'required': _("Operation wasn't completed.")},
                                   widget=forms.HiddenInput)
+    user_handle = forms.CharField(widget=forms.HiddenInput)
```
Comment (tpazderka) [inline django_fido/forms.py:22]: This field should not be required or possibly dependent on `resident_key` settings.

**Review by tpazderka — CHANGES_REQUESTED** [review]

File: django_fido/forms.py (hunk 1 above)
Comment (variable) [reply django_fido/forms.py:22]: Initially I had this field not required and it seemed dependent on the `resident_key`, but I realised this not the case, user id is always present during registration process, the `resident_key` merely telling the yubikey to store the credential on the key. 

In short, both features are required to make passwordless to work but they are independent on their own.

File: .idea/workspace.xml [hunk 2]
```diff
@@ -0,0 +1,80 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tpazderka) [inline .idea/workspace.xml:1]: This file snuck back in :)

File: django_fido/forms.py (hunk 1 above)
Comment (tpazderka) [reply django_fido/forms.py:22]: OK, I had no idea about that.

**Review by tpazderka — CHANGES_REQUESTED** [review]
Again the `.idea` file.

File: .idea/workspace.xml (hunk 2 above)
Comment (variable) [reply .idea/workspace.xml]: argggghhhh, removed again.

**Comment by ziima** [comment]
> Nevermind, found the issue, `two_setp_auth` in the test settings is `True`, but my project settings has set it to `False`

When something like this happens, it usually means that tests are not properly isolated. It would be better to set the setting in tests correctly, so they are independent on the settings actually used. Not in this PR though.

**Review by tpazderka — APPROVED** [review]

**Review by ziima — APPROVED** [review]

**Comment by variable** [comment]
Hmmmmmmm, I just realised this is not going to fly when we run the migration for existing users, because the user_handle field is supposed to be unique, even we work out a magical way to populate some random string for each existing user, the passwordless won't match the random string until user re-register the key. Doh....

**Comment by variable** [comment]
> Hmmmmmmm, I just realised this is not going to fly when we run the migration for existing users, because the user_handle field is supposed to be unique, even we work out a magical way to populate some random string for each existing user, the passwordless won't match the random string until user re-register the key. Doh....

Oh wait, is it a false positive? Null is unique at least to postgres.

**Comment by variable** [comment]
I guess one way to fix this is to make the new `user_handle` field nullable, I know it's against the django recommendation not to set text fields to null...

**Comment by variable** [comment]
Hello guys,

Please rip out this PR from master, I just came to realised the user handle can be decoded into username by using `base64.b64decode(value)` so this field is not needed since we can simply using the username from the userhandle provided by `resident_key` option to find the user.

**Comment by variable** [comment]
> Hello guys,
> 
> Please rip out this PR from master, I just came to realised the user handle can be decoded into username by using `base64.b64decode(value)` so this field is not needed since we can simply using the username from the userhandle provided by `resident_key` option to find the user.

Please ignore this as we should not use username as user handle when we have resident_key=True

**Comment by tpazderka** [comment]
But we still need to fix the migration for existing users.

**Comment by variable** [comment]
> But we still need to fix the migration for existing users.

Yes, there is another PR on top of this which removes the `unique=True` constraint.

**Pull request closed without merging**